README
NAME
    IPC::Run::Fused - Capture Stdout/Stderr simultaneously as if it were one
    stream, painlessly.

VERSION
    version 1.000001

SYNOPSIS
      use IPC::Run::Fused qw( run_fused );

      run_fused( my $fh, $stderror_filled_program, '--file', $tricky_filename, @moreargs ) || die "Argh $@";
      open my $fh, '>', 'somefile.txt' || die "NOO  $@";

      # Simple implementation of 'tee' like behaviour,
      # sending to stdout and to a file.

      while ( my $line = <$fh> ) {
        print $fh $line;
        print $line;
      }

DESCRIPTION
    Have you ever tried to do essentially the same as you would in bash do
    to this:

      parentapp <( app 2>&1  )

    And found massive road works getting in the way.

    Sure, you can always do this style syntax:

      open my $fh, 'someapp --args foo 2>&1 |';

    But that's not very nice, because

    1. you're relying on a sub-shell to do that for you
    2. you have to manually escape everything
    3. you can't use list context.

    And none of this is very Modern or Nice

SIMPLEST THING THAT JUST WORKS
    This code is barely tested, its here, because I spent hours griping
    about how the existing ways suck.

FEATURES
    1. No String Interpolation.
        Arguments after the first work as if you'd passed them directly to
        'system'. You can be as dangerous or as safe as you want with them.
        We recommend passing a list, but a string ( as a scalar reference )
        should work

        But if you're using a string, this modules probably not affording
        you much.

    2. No dicking around with managing multiple file handles yourself.
        I looked at "IPC::Run", "IPC::Run3" and "IPC::Open3", and they all
        seemed very unfriendly, and none did what I wanted.

    3. Non-global file-handles supported by design.
        All the competition seem to still have this thing for global file
        handles and you having to use them. Yuck!.

        We have a few global file-handles inside our code, but they're only
        "STDERR" and "STDOUT", at present I don't think I can circumvent
        that. If I ever can, I'll endeavor to do so ☺

EXPORTED FUNCTIONS
    No functions are exported by default, be explicit with what you want.

    At this time there is only one, and there are no plans for more.

  run_fused
      run_fused( $fh, $executable, @params ) || die "$@";
      run_fused( $fh, \$command_string )     || die "$@";
      run_fused( $fh, sub { .. } )           || die "$@";

      # Recommended

      run_fused( my $fh, $execuable, @params ) || die "$@";

      # Somewhat supported

      run_fused( my $fh, \$command_string ) || die "$@";

    $fh will be clobbered like 'open' does, and $cmd, @args will be passed,
    as-is, through to exec() or system().

    $fh will point to an IO::Handle attached to the end of a pipe running
    back to the called application.

    the command will be run in a fork, and "STDERR" and "STDOUT" "fused"
    into a singular pipe.

    NOTE: at present, "STDIN"'s file-descriptor is left unchanged, and child
    processes will inherit parent "STDIN"'s, and will thus block ( somewhere
    ) waiting for response.

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for IPC-Run-Fused

1.000002 2017-03-08T15:25:49Z 72125a3

1.000001 2017-03-08T14:46:47Z 9a6d4ee
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +14 ↑4 -3 (recommends: +1, suggests: ↑2)
 - test: +1 ↓1 (recommends: ↑1 ↓1)

1.000000 2014-10-03T05:04:57Z
 [00 - Important - Version Scheme Change]
 - Using 1.xxxyyy ( mantissa=6 ) instead of 0.xxxyyyzz
 - This may affect hand-operated downstreams
 - This is the primary reason for the 1.x on the box.

 [01 - Important - Toolkit Switch]
 - Tooling now uses ExtUtils::MakeMaker
 - This may affect hand-operated downstreams
 - "Latest" Testing dependencies now merely recommendations.
 - https://questhub.io/realm/perl/quest/5388cc12bbd0be5b2500011f

 [Dependencies::Stats]
 - Dependencies changed since 0.04000100, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +46 -2 (recommends: -1, suggests: +1 ↑1)
 - runtime: +2 -1 (suggests: -2)
 - test: +2 ↓1 -2 (recommends: +3)

 [Internals]
 - use lower level Exporter code instead of Sub::Exporter
 - Avoid need for builder method and bake implementation at USE
 - Newer technique for passing vivifable references around internally employed to make parameter passing more apparent.

 [Metadata]
 - Specify email address in copyright data
 - Migrate repo to kentnl org.

 [Packaging]
 - Smarter Handling of Win32 Prereqs

0.04000100 2013-02-05T07:17:18Z
 [BugFix]
 - load IO::Handle in the test script.

0.04000000 2013-02-05T05:14:00Z
 [!Features]
 - Should now actually work on Win32 due to the discovery of Win32::Job
 - Fail behaviour should be more concise with additional context.

 [BugFix]
 - previous release accidentally dropped _fail method.

 [Dependencies::Stats]
 - Dependencies changed since 0.03000000, see misc/*.deps* for details
 - runtime: +2 ↑1 (suggests: +2)

 [Internals]
 - _fail method now unified via goto \&IPC::Run::Fused::_fail
 - new ::Win32 namespace to implement Win32 Features

 [Tests]
 - reworked to not be order dependent, only flush-atomic.
 - previously disabled tests re-enabled on Win32

0.03000000 2013-02-03T00:47:09Z
 [!Features]
 - Most the code for Unix-like environments moved to ::POSIX with a proxy method in the main namespace that will defer
   to the right backend depending on the value of $^O.
 - This serves as a stepping stone to working Win32 Support.
 - Win32 Currently causes a die() because its broken anyway atm.
 - Added support for passing a coderef to run.
 - IO::Pipe code dropped and just uses a plain `pipe` which is less magical.

 [Dependencies::Stats]
 - Dependencies changed since 0.02000000, see misc/*.deps* for details
 - runtime: ↓1 -2

 [Modules::New]
 - IPC::Run::Fused::POSIX

 [Tests::New]
 - t/03-captures-nodelay-fork.t

0.02000000 2013-01-26T01:07:56Z
 - Guts removed mostly and replaced with IO::Pipe, because the existing code was broken and not very portable.

 [Dependencies::Stats]
 - Dependencies changed since 0.01028807, see misc/*.deps* for details
 - runtime: +1

0.01028807 2013-01-26T00:26:12Z
 - Maint release.

 [BugFixes]
 - disabled tests on Win32. We have had it work on win32, but the tests are broken on win32, and we're not quite sure
   why yet. Code is planned to have a total overhaul at some point.

 [Dependencies::Stats]
 - Dependencies changed since 0.01028806, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +6 (recommends: +1, suggests: +1)
 - runtime: +2 ↑1
 - test: ↑1

 [Documentation]
 - add README.mkdn to dist
 - normalise Changes to CPAN::Changes form
 - update copyright license year
 - reindent LICENSE file
 - update fsf address

 [Internals]
 - declare $AUTHORITY
 - change $VERSION declaration form

 [Meta]
 - set bug tracker to github issues
 - use https:// for all git uris.
 - add x_authority

 [Packaging]
 - move to @Author::KENTNL
 - move author/release tests to xt/ and stay there

 [Tests]
 - regenerate 00-compile.t
 - regenerate 000-report-versions-tiny.t

0.01028806 2010-07-30T05:52:27Z
 - Migrate code to [@KENTNL]
 - Fix non-portable vstring in use $version;

 [Dependencies::Stats]
 - Dependencies changed since 0.01001607, see misc/*.deps* for details
 - build: +1
 - configure: +1
 - runtime: +1 -2
 - test: +4

0.01001607 2009-11-06T06:34:07Z
 - Native pipes by default.
 - Removed needless POSIX::dup();
 - Loads more error handling to try work out whats going wrong in tests.
 - Reorgansied fork() code into a handy subroutine.
 - Reduced Test-grinding from 400 to 100, as 100 seems enough to make problems turn up if there are any

0.01001122 2009-11-01T21:22:16Z
 - Fall back to native pipes if POSIX pipes fail. Experimental, so will carp.

0.01000514 2009-10-26T13:46:20Z
 - Hard depends on perl 5.8.0 or greater due to >>&= File Open
 - Fixed up the absense of parentheses in synopsis and tests.
 - Tidied Synopsis Code.
 - run_fused( $fh, 'Command String | bar' ) will now be interpreted as
 - if to look for a literal program sharing that literal name. To get the prior behaviour,
 - pass a scalar ref instead, ie: run_fused( $fh, \'Command String' ).
 - run_fused( $fh, @list ) now properly calls exec with exec { $program } @args
 - dropped Time::HiRes and now use a stupid arthimetic loop to delay things.

 [Dependencies::Stats]
 - Dependencies changed since 0.01000001, see misc/*.deps* for details
 - runtime: -1

0.01000001 2009-10-21T00:36:09Z
 - First version.


t/tbin/01.pl
#!/usr/bin/perl
use strict;
use warnings;

*STDOUT->autoflush(0);
*STDERR->autoflush(0);

for my $m ( 0 .. 5 ) {
  print "<<$m>>\n";
  for ( 0 .. 10 ) {
    print {*STDERR} "w$_";
  }
  print {*STDERR} "\n";
  for ( 0 .. 10 ) {
    print {*STDOUT} "p$_";
  }
  print {*STDOUT} "\n";
}


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4210"
  },
  "configure_requires" => {
    "Module::Build" => "0.4210"
  },
  "dist_abstract" => "Capture Stdout/Stderr simultaneously as if it were one stream, painlessly.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "IPC-Run-Fused",
  "dist_version" => "0.04000101",
  "license" => "perl",
  "module_name" => "IPC::Run::Fused",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "IO::Handle" => 0,
    "Module::Runtime" => 0,
    "Sub::Exporter" => "0.982",
    "perl" => "v5.8.3",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "Test::More" => "1.001006"
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "Module::Build" => "0.4210",
  "Test::More" => "1.001006"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/IPC/Run/Fused.pm
use 5.008003;
use strict;
use warnings;

package IPC::Run::Fused;

our $VERSION = '1.000002';

# ABSTRACT: Capture Stdout/Stderr simultaneously as if it were one stream, painlessly.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY







































































































our %FAIL_CONTEXT;

use Exporter qw(import);
sub run_fused;
our @EXPORT_OK = qw( run_fused _fail );

sub _stringify {
  my ($entry) = @_;
  return 'undef' if not defined $entry;
  return qq{`$entry`};
}

sub _fail {    ## no critic (Subroutines::RequireArgUnpacking)
  my (@message) = @_;
  my $errors = {
    ## no critic (ValuesAndExpressions::RequireInterpolationOfMetachars)
    q[$?]  => _stringify($?),
    q[$!]  => _stringify($!),
    q[$^E] => _stringify($^E),
    q[$@]  => _stringify($@),
    ( map { $_ => _stringify( $FAIL_CONTEXT{$_} ) } keys %FAIL_CONTEXT ),
  };
  my $message = q[];
  $message .= qq{\n} . $_ for @message;
  $message .= sprintf qq{\n\t%s => %s}, $_, $errors->{$_} for sort keys %{$errors};
  $message .= qq{\n};
  require Carp;
  @_ = $message;
  goto \&Carp::confess;
}

sub _win32_run_fused {
  require IPC::Run::Fused::Win32;
  goto \&IPC::Run::Fused::Win32::run_fused;
}

sub _posix_run_fused {
  require IPC::Run::Fused::POSIX;
  goto \&IPC::Run::Fused::POSIX::run_fused;
}

if ( 'MSWin32' eq $^O ) {
  *run_fused = \&_win32_run_fused;
}
else {
  *run_fused = \&_posix_run_fused;
}
1;

__END__

=pod

=encoding UTF-8

=head1 NAME

IPC::Run::Fused - Capture Stdout/Stderr simultaneously as if it were one stream, painlessly.

=head1 VERSION

version 1.000002

=head1 SYNOPSIS

  use IPC::Run::Fused qw( run_fused );

  run_fused( my $fh, $stderror_filled_program, '--file', $tricky_filename, @moreargs ) || die "Argh $@";
  open my $fh, '>', 'somefile.txt' || die "NOO  $@";

  # Simple implementation of 'tee' like behaviour,
  # sending to stdout and to a file.

  while ( my $line = <$fh> ) {
    print $fh $line;
    print $line;
  }

=head1 DESCRIPTION

Have you ever tried to do essentially the same as you would in bash do to this:

  parentapp <( app 2>&1  )

And found massive road works getting in the way.

Sure, you can always do this style syntax:

  open my $fh, 'someapp --args foo 2>&1 |';

But that's not very nice, because

=over 4

=item 1. you're relying on a sub-shell to do that for you

=item 2. you have to manually escape everything

=item 3. you can't use list context.

=back

And none of this is very B<Modern> or B<Nice>

=head1 SIMPLEST THING THAT JUST WORKS

This code is barely tested, its here, because I spent hours griping about how the existing ways suck.

=head1 FEATURES

=over 4

=item 1. No String Interpolation.

Arguments after the first work as if you'd passed them directly to 'system'. You can be as dangerous or as
safe as you want with them. We recommend passing a list, but a string ( as a scalar reference ) should work

But if you're using a string, this modules probably not affording you much.

=item 2. No dicking around with managing multiple file handles yourself.

I looked at L<< C<IPC::Run>|IPC::Run >>, L<< C<IPC::Run3>|IPC::Run3 >> and L<< C<IPC::Open3>|IPC::Open3 >>, and they all seemed
very unfriendly, and none did what I wanted.

=item 3. Non-global file-handles supported by design.

All the competition seem to still have this thing for global file handles and you having to use them. Yuck!.

We have a few global file-handles inside our code, but they're only C<STDERR> and C<STDOUT>, at present I don't think I can
circumvent that. If I ever can, I'll endeavor to do so ☺

=back

=head1 EXPORTED FUNCTIONS

No functions are exported by default, be explicit with what you want.

At this time there is only one, and there are no plans for more.

=head2 run_fused

  run_fused( $fh, $executable, @params ) || die "$@";
  run_fused( $fh, \$command_string )     || die "$@";
  run_fused( $fh, sub { .. } )           || die "$@";

  # Recommended

  run_fused( my $fh, $execuable, @params ) || die "$@";

  # Somewhat supported

  run_fused( my $fh, \$command_string ) || die "$@";

$fh will be clobbered like 'open' does, and $cmd, @args will be passed, as-is, through to exec() or system().

$fh will point to an IO::Handle attached to the end of a pipe running back to the called application.

the command will be run in a fork, and C<STDERR> and C<STDOUT> "fused" into a singular pipe.

B<NOTE:> at present, C<STDIN>'s file-descriptor is left unchanged, and child processes will inherit parent C<STDIN>'s, and will thus block ( somewhere ) waiting for response.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/IPC/Run/Fused/POSIX.pm
use 5.008003;
use strict;
use warnings;

package IPC::Run::Fused::POSIX;

our $VERSION = '1.000002';

# ABSTRACT: Implementation of IPC::Run::Fused for POSIX-ish systems.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use IO::Handle;

























use Exporter qw(import);
use IPC::Run::Fused qw( _fail );

our @EXPORT_OK = qw( run_fused );

sub run_fused {    ## no critic ( Subroutines::RequireArgUnpacking )

  my ( $read_handle, @params ) = ( \shift @_, @_ );

  my ( $reader, $writer );

  pipe $reader, $writer or _fail('Creating Pipe');

  if ( my $pid = fork ) {
    ${$read_handle} = $reader;
    return $pid;
  }

  open *STDOUT, '>>&=', $writer->fileno or _fail('Assigning to STDOUT');
  open *STDERR, '>>&=', $writer->fileno or _fail('Assigning to STDERR');

  if ( ref $params[0] and 'CODE' eq ref $params[0] ) {
    $params[0]->();
    exit;
  }
  if ( ref $params[0] and 'SCALAR' eq ref $params[0] ) {
    my $command = ${ $params[0] };
    exec $command or _fail('<<exec command>> failed');
    exit;
  }

  my $program = $params[0];
  ## no critic (ValuesAndExpressions::RequireInterpolationOfMetachars)
  exec {$program} @params or _fail('<<exec {program} @argv>> failed');
  exit;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

IPC::Run::Fused::POSIX - Implementation of IPC::Run::Fused for POSIX-ish systems.

=head1 VERSION

version 1.000002

=head1 METHODS

=head2 run_fused

  run_fused( $fh, $executable, @params ) || die "$@";
  run_fused( $fh, \$command_string )     || die "$@";
  run_fused( $fh, sub { .. } )           || die "$@";

  # Recommended

  run_fused( my $fh, $execuable, @params ) || die "$@";

  # Somewhat supported

  run_fused( my $fh, \$command_string ) || die "$@";

$fh will be clobbered like 'open' does, and $cmd, @args will be passed, as-is, through to exec() or system().

$fh will point to an IO::Handle attached to the end of a pipe running back to the called application.

the command will be run in a fork, and C<STDERR> and C<STDOUT> "fused" into a singular pipe.

B<NOTE:> at present, C<STDIN>'s FD is left unchanged, and child processes will inherit parent C<STDIN>'s, and will thus block ( somewhere ) waiting for response.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/IPC/Run/Fused/Win32.pm
use 5.008003;
use strict;
use warnings;

package IPC::Run::Fused::Win32;

our $VERSION = '1.000002';

# ABSTRACT: Implementation of IPC::Run::Fused for Win32

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use IO::Handle;
use Module::Runtime;


























use IPC::Run::Fused qw(_fail);
use Socket qw( AF_UNIX SOCK_STREAM PF_UNSPEC );

use Exporter qw(import);
our @EXPORT_OK = qw( run_fused );

sub run_fused {
  my ( undef, @params ) = @_;
  if ( ref $params[0] and 'CODE' eq ref $params[0] ) {
    goto \&_run_fused_coderef;
  }
  goto \&_run_fused_job;
}

sub _run_fused_job {    ## no critic (Subroutines::RequireArgUnpacking)
  my ( $read_handle, @params ) = ( \shift @_, @_ );

  my $config = _run_fused_jobdecode(@params);

  Module::Runtime::require_module('File::Which');

  $config->{which} = File::Which::which( $config->{executable} );

  local $IPC::Run::Fused::FAIL_CONTEXT{which}      = $config->{which};
  local $IPC::Run::Fused::FAIL_CONTEXT{executable} = $config->{executable};
  local $IPC::Run::Fused::FAIL_CONTEXT{command}    = $config->{command};

  if ( not $config->{which} ) {
    _fail('Failed to resolve executable to path');
  }

  Module::Runtime::require_module('Win32::Job');

  pipe ${$read_handle}, my $writer;

  if ( my $pid = fork ) {
    return $pid;
  }

  my $job = Win32::Job->new();
  $job->spawn(
    $config->{which},
    $config->{command},
    {
      stdout => $writer,
      stderr => $writer,
    },
  ) or _fail('Could not spawn job');
  my $result = $job->run( -1, 0 );
  if ( not $result ) {
    my $status = $job->status();
    if ( exists $status->{exitcode} and 293 == $status->{exitcode} ) {
      _fail('Process used more than allotted time');
    }
    _fail( 'Child process terminated with exit code' . $status->{exitcode} );
  }
  exit;
}

sub _run_fused_jobdecode {
  my (@params) = @_;

  if ( ref $params[0] and 'SCALAR' eq ref $params[0] ) {
    my $command = ${ $params[0] };
    $command =~ s/\A\s*//msx;
    return {
      command    => $command,
      executable => _win32_command_find_invocant($command),
    };
  }
  return {
    executable => $params[0],
    command    => _win32_escape_command(@params),
  };
}

sub _run_fused_coderef {    ## no critic (Subroutines::RequireArgUnpacking)
  my ( $read_handle, $code ) = ( \shift @_, @_ );
  my ( $reader, $writer );

  socketpair $reader, $writer, AF_UNIX, SOCK_STREAM, PF_UNSPEC or _fail('creating socketpair');
  shutdown $reader, 1 or _fail('Cant close write to reader');
  shutdown $writer, 0 or _fail('Cant close read to writer');

  if ( my $pid = fork ) {
    ${$read_handle} = $reader;
    return $pid;
  }

  close *STDERR or _fail('Closing STDERR');
  close *STDOUT or _fail('Closing STDOUT');
  open *STDOUT, '>>&=', $writer or _fail('Assigning to STDOUT');
  open *STDERR, '>>&=', $writer or _fail('Assigning to STDERR');
  $code->();
  exit;

}

our $BACKSLASH         = chr 92;
our $DBLBACKSLASH      = $BACKSLASH x 2;
our $DOS_SPECIAL_CHARS = {
  chr 92 => [ 'backslash ',    $BACKSLASH x 2 ],
  chr 34 => [ 'double-quotes', $BACKSLASH . chr 34 ],

  #chr(60) => ['open angle bracket', $backslash . chr(60)],
  #chr(62) => ['close angle bracket', $backslash . chr(62)],
};
our $DOS_REV_CHARS = {
  map { ( $DOS_SPECIAL_CHARS->{$_}->[1], [ $DOS_SPECIAL_CHARS->{$_}->[0], $_ ] ) }
    keys %{$DOS_SPECIAL_CHARS},
};

sub _win32_escape_command_char {
  my ($char) = @_;
  return $char unless exists $DOS_SPECIAL_CHARS->{$char};
  return $DOS_SPECIAL_CHARS->{$char}->[1];
}

sub _win32_escape_command_token {
  ## no critic (RegularExpressions)
  my $chars = join q{}, map { _win32_escape_command_char($_) } split //, shift;
  return qq{"$chars"};
}

sub _win32_escape_command {
  my (@tokens) = @_;
  return join q{ }, map { _win32_escape_command_token($_) } @tokens;
}

sub _win32_command_find_invocant {
  my ($command) = @_;
  $command = "$command";
  my $first = q[];
  ## no critic (RegularExpressions)
  my @chars = split //, $command;
  my $inquote;

  while (@chars) {
    my $char  = $chars[0];
    my $dchar = $chars[0] . $chars[1];

    if ( not $inquote and q["] eq $char ) {
      $inquote = 1;
      shift @chars;
      next;
    }
    if ( $inquote and q["] eq $char ) {
      $inquote = undef;
      shift @chars;
      next;
    }
    if ( exists $DOS_REV_CHARS->{$dchar} ) {
      $first .= $DOS_REV_CHARS->{$dchar}->[1];
      shift @chars;
      shift @chars;
      next;
    }
    if ( q[ ] eq $char and not $inquote ) {
      if ( not length $first ) {
        shift @chars;
        next;
      }
      return $first;
    }
    if ( q[ ] eq $char and $inquote ) {
      $first .= $char;
      shift @chars;
      next;
    }
    $first .= $char;
    shift @chars;
  }
  if ($inquote) {
    _fail('Could not parse command from commandline');
  }
  return $first;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

IPC::Run::Fused::Win32 - Implementation of IPC::Run::Fused for Win32

=head1 VERSION

version 1.000002

=head1 METHODS

=head2 run_fused

  run_fused( $fh, $executable, @params ) || die "$@";
  run_fused( $fh, \$command_string )     || die "$@";
  run_fused( $fh, sub { .. } )           || die "$@";

  # Recommended

  run_fused( my $fh, $executable, @params ) || die "$@";

  # Somewhat supported

  run_fused( my $fh, \$command_string ) || die "$@";

$fh will be clobbered like 'open' does, and $cmd, @args will be passed, as-is, through to exec() or system().

$fh will point to an IO::Handle attached to the end of a pipe running back to the called application.

the command will be run in a fork, and C<STDERR> and C<STDOUT> "fused" into a singular pipe.

B<NOTE:> at present, C<STDIN>'s FD is left unchanged, and child processes will inherit parent C<STDIN>'s, and will thus block
( somewhere ) waiting for response.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.008003;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Capture Stdout/Stderr simultaneously as if it were one stream, painlessly.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "IPC-Run-Fused",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008003",
  "NAME" => "IPC::Run::Fused",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Exporter" => 0,
    "IO::Handle" => 0,
    "Module::Runtime" => 0,
    "Socket" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "Test::More" => 0
  },
  "VERSION" => "1.000002",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "IO::Handle" => 0,
  "Module::Runtime" => 0,
  "Socket" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

if ( $^O eq 'MSWin32' ) {
	$WriteMakefileArgs{PREREQ_PM}{'File::Which'} = '0';
	$WriteMakefileArgs{PREREQ_PM}{'Win32::Job'} = '0';
}

WriteMakefile(%WriteMakefileArgs);
